File: Python/209__Minimum_Size_Subarray_Sum.py
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        if sum(nums) < s:
            return 0

        start, end = 0, 0
        curval = nums[0]
        result = len(nums)
        while end != len(nums) - 1:
            while end < len(nums) - 1 and curval < s:
                end += 1
                curval += nums[end]
            while start < end and curval - nums[start] >= s:
                curval -= nums[start]
                start += 1
            result = min(result, end - start + 1)
            curval -= nums[start]
            start += 1
        return result

# Same Approach But Faster and More Clear Solution

class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        val = left = 0
        result = len(nums) + 1
        for right, n in enumerate(nums):
            val += n
            while val >= s:
                result = min(result, right - left + 1)
                val -= nums[left]
                left += 1
        return result if result <= len(nums) else 0

File: Java/209__Minimum_Size_Subarray_Sum.java
class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int sum = 0, l = 0;;
        int result = Integer.MAX_VALUE;
        for (int r = 0; r < nums.length; r++) {
            sum += nums[r];
            if (sum < s) continue;
            while (sum >= s) {
                sum -= nums[l];
                l++;
            }
            result = Math.min(result, r - l + 2);
        }
        return result == Integer.MAX_VALUE ? 0 : result;
    }
}
